main/src/ca/uwaterloo/flix/language/phase/unification/CaseSetUnification2.scala

```diff
@@ -0,0 +1,547 @@
+
+    // ¬(¬x ∨ y) => x ∧ ¬y
+    case SetFormula.Or(SetFormula.Not(x), y) =>
+      mkIntersection(x, mkComplement(y, univ), univ)
+
+    // ¬(x ∨ ¬y) => ¬x ∧ y
+    case SetFormula.Or(x, SetFormula.Not(y)) =>
+      mkIntersection(mkComplement(x, univ), y, univ)
+
+    case _ => SetFormula.Not(tpe0)
+  }
+
+  /**
+    * Returns the conjunction of the two set formulas `tpe1` and `tpe2`.
+    */
+  // NB: The order of cases has been determined by code coverage analysis.
+  @tailrec
+  def mkIntersection(tpe1: SetFormula, tpe2: SetFormula, univ: Set[Int]): SetFormula = (tpe1, tpe2) match {
+    case (SetFormula.Cst(x1), SetFormula.Cst(x2)) =>
+      SetFormula.Cst(x1 & x2)
+
+    // ¬x ∧ (x ∨ y) => ¬x ∧ y
+    case (SetFormula.Not(x1), SetFormula.Or(x2, y)) if x1 == x2 =>
+      mkIntersection(mkComplement(x1, univ), y, univ)
+
+    // x ∧ ¬x => F
+    case (x1, SetFormula.Not(x2)) if x1 == x2 =>
+      SetFormula.Empty
+
+    // ¬x ∧ x => F
+    case (SetFormula.Not(x1), x2) if x1 == x2 =>
+      SetFormula.Empty
+
+    // x ∧ (x ∧ y) => (x ∧ y)
+    case (x1, SetFormula.And(x2, y)) if x1 == x2 =>
+      mkIntersection(x1, y, univ)
+
+    // x ∧ (y ∧ x) => (x ∧ y)
+    case (x1, SetFormula.And(y, x2)) if x1 == x2 =>
+      mkIntersection(x1, y, univ)
+
+    // (x ∧ y) ∧ x) => (x ∧ y)
+    case (SetFormula.And(x1, y), x2) if x1 == x2 =>
+      mkIntersection(x1, y, univ)
+
+    // (x ∧ y) ∧ y) => (x ∧ y)
+    case (SetFormula.And(x, y1), y2) if y1 == y2 =>
+      mkIntersection(x, y1, univ)
+
+    // x ∧ (x ∨ y) => x
+    case (x1, SetFormula.Or(x2, _)) if x1 == x2 =>
+      x1
+
+    // (x ∨ y) ∧ x => x
+    case (SetFormula.Or(x1, _), x2) if x1 == x2 =>
+      x1
+
+    // x ∧ (y ∧ ¬x) => F
+    case (x1, SetFormula.And(_, SetFormula.Not(x2))) if x1 == x2 =>
+      SetFormula.Empty
+
+    // (¬x ∧ y) ∧ x => F
+    case (SetFormula.And(SetFormula.Not(x1), _), x2) if x1 == x2 =>
+      SetFormula.Empty
+
+    // x ∧ ¬(x ∨ y) => F
+    case (x1, SetFormula.Not(SetFormula.Or(x2, _))) if x1 == x2 =>
+      SetFormula.Empty
+
+    // ¬(x ∨ y) ∧ x => F
+    case (SetFormula.Not(SetFormula.Or(x1, _)), x2) if x1 == x2 =>
+      SetFormula.Empty
+
+    // x ∧ (¬x ∧ y) => F
+    case (x1, SetFormula.And(SetFormula.Not(x2), _)) if x1 == x2 =>
+      SetFormula.Empty
+
+    // (¬x ∧ y) ∧ x => F
+    case (SetFormula.And(SetFormula.Not(x1), _), x2) if x1 == x2 =>
+      SetFormula.Empty
+
+    // x ∧ x => x
+    case _ if tpe1 == tpe2 => tpe1
+
+    case _ =>
+      //      val s = s"And($eff1, $eff2)"
+      //      val len = s.length
+      //      if (true) {
+      //        println(s.substring(0, Math.min(len, 300)))
+      //      }
+
+      SetFormula.And(tpe1, tpe2)
+  }
+
+  /**
+    * Returns the disjunction of the two set formulas `tpe1` and `tpe2`.
+    */
+  // NB: The order of cases has been determined by code coverage analysis.
+  @tailrec
+  def mkUnion(tpe1: SetFormula, tpe2: SetFormula, univ: Set[Int]): SetFormula = (tpe1, tpe2) match {
```

Comment:
These are better/different than the ones in SetFormulaAlg. Could we combine them and have them once or is there are reason to have them twice?